x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/DeprecationRoleDescriptorConsumer.java

```diff
@@ -0,0 +1,158 @@
+ * each role descriptor and alias pair, and it contains all the indices for which privileges are a subset of those of the alias. In this
+ * case, the administrator has to adjust the index privileges definition of the respective role such that name patterns do not cover aliases
+ * (or rename aliases). If no logging is generated then the roles used for the current indices and aliases are not vulnerable to the
+ * subsequent breaking change. However, there could be role descriptors that are not used (not mapped to a user that is currently using the
+ * system) which are invisible to this check. Moreover, role descriptors can be dynamically added by role providers. In addition, role
+ * descriptors are merged when building the effective role, so a role-alias pair reported as deprecated might not actually have an impact if
+ * other role descriptors cover its indices. The check iterates over all indices and aliases for each role descriptor so it is quite
+ * expensive computationally. For this reason the check is done only once a day for each role. If the role definitions stay the same, the
+ * deprecations can change from one day to another only if aliases or indices are added.
+ */
+public final class DeprecationRoleDescriptorConsumer implements Consumer<Collection<RoleDescriptor>> {
+
+    private static final String ROLE_PERMISSION_DEPRECATION_STANZA = "Role [%s] contains index privileges covering the [%s] alias but"
+            + " which do not cover some of the indices that it points to [%s]. Granting privileges over an alias and hence granting"
+            + " privileges over all the indices that the alias points to is deprecated and will be removed in a future version of"
+            + " Elasticsearch. Instead define permissions exclusively on index names or index name patterns.";
+
+    private final DeprecationLogger deprecationLogger;
+    private final ClusterService clusterService;
+    private final ThreadPool threadPool;
+    private final Set<String> dailyRoleCache;
+
+    public DeprecationRoleDescriptorConsumer(ClusterService clusterService, ThreadPool threadPool, DeprecationLogger deprecationLogger) {
+        this.deprecationLogger = deprecationLogger;
+        this.clusterService = clusterService;
+        this.threadPool = threadPool;
+        // this String Set keeps "<date>-<role>" pairs so that we only log a role once a day.
+        this.dailyRoleCache = Collections.newSetFromMap(Collections.synchronizedMap(new LinkedHashMap<String, Boolean>() {
+            @Override
+            protected boolean removeEldestEntry(final Map.Entry<String, Boolean> eldest) {
+                return false == eldest.getKey().startsWith(todayISODate());
+            }
+        }));
+    }
+
+    @Override
+    public void accept(Collection<RoleDescriptor> effectiveRoleDescriptors) {
+        threadPool.generic().execute(() -> {
```

Comment:
I wonder if we should bound the concurrent running of these items? If it is computationally expensive, maybe it is best to run one at a time?

Reply:
This is a good point as well. In the worst case we might require one generic thread pool worker for each role descriptor.

I have pushed 6ed6665d8c313761c0c316e7eb that makes sure only **one Role Descriptor** is processed at a time. Moreover, a maximum of **one thread pool worker** is recruited at a time.
Each submitted `Runnable` tries to process as many Role Descriptors as it can, by popping a synchronized work queue. A new `Runnable` is created (synchronously) only if the work queue is empty - so there is no other `Runnable` working on the queue's head item.